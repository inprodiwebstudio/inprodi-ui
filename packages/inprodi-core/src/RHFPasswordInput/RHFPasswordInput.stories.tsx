import React                                       from "react";
import { useForm, FormProvider }                   from "react-hook-form";
import { ComponentMeta, ComponentStory }           from "@storybook/react";
import { RHFPasswordInput, RHFPasswordInputProps } from "./RHFPasswordInput";

export default {
	title     : "Core/Form/PasswordInput",
	component : RHFPasswordInput,
	argTypes  : {
		size   : { control : "select" },
		radius : { control : "select" },
	},
} as ComponentMeta<typeof RHFPasswordInput>;

const Template: ComponentStory<typeof RHFPasswordInput> = (args : RHFPasswordInputProps) => {
	const methods = useForm();
	return(
		<FormProvider {...methods}>
			<RHFPasswordInput {...args} />
		</FormProvider>
	);
};

export const Basic = Template.bind({});
Basic.args = {
	label        : "Password",
	name         : "basic-inpt-password",
	description  : "",
	error        : false,
	radius       : "sm",
	size         : "md",
	withAsterisk : false,
};
export const Styled = Template.bind({});
Styled.args = {
	label        : "Password",
	name         : "basic-inpt-password",
	description  : "description",
	error        : false,
	radius       : "md",
	size         : "lg",
	withAsterisk : true,
};
export const Invalid = Template.bind({});
Invalid.args = {
	label        : "Password",
	name         : "basic-inpt-password",
	description  : "",
	error        : true,
	radius       : "sm",
	size         : "md",
	withAsterisk : false,
};

