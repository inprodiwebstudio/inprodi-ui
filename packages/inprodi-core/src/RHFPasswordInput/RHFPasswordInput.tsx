import React                                              from "react";
import { useFormContext, Controller }                     from "react-hook-form";
import { MantineSize, PasswordInput, PasswordInputProps } from "@mantine/core";
import { useFocusWithin }                                 from "@mantine/hooks";
import { OpenedEyeOutline, ClosedEyeOutline }             from "../../../inprodi-icons/dist";
import styles                                             from "../styles";

export interface RHFPasswordInputProps extends PasswordInputProps, React.RefAttributes<HTMLInputElement> {
    /**
    * Input's name being registered. [controller](https://react-hook-form.com/api/usecontroller/controller)
    */
	name : string;
    /**
    * Input's label
    */
	label ?: string;
    /**
    * An extra action to invoke when the inpt changes. Note: the library still control the input's value.
    */
	onChange ?: (event : React.ChangeEvent<HTMLInputElement>) => void;
    /**
    * Input description, displayed after label [PasswordInput](https://mantine.dev/core/password-input/)
    */
    description ?: string;
    /**
    * Determines whether required asterisk should be rendered, overrides required prop, does not add required attribute to the input [PasswordInput](https://mantine.dev/core/password-input/)
    */
    withAsterisk ?: boolean;
    /**
    * to change the inpt size.To see more: [textInput](https://mantine.dev/core/text-input/)
    */
   size?: MantineSize;
	/**
    * to change the inpt radius.To see more: [textInput](https://mantine.dev/core/text-input/)
    */
    radius?: MantineSize;
}
export const RHFPasswordInput = ({
	name,
	label,
	onChange,
	...rest
} : RHFPasswordInputProps) => {
	const { ref, focused } = useFocusWithin();
	const { classes } = styles({ focused });
	const { control } = useFormContext();

	const getVisibilityIcon = ({ reveal, size } : { reveal : boolean; size : number }) =>{
		return reveal ? <OpenedEyeOutline data-testid="opened-eye" fontSize={size} />
			: <ClosedEyeOutline data-testid="closed-eye" fontSize={size} />;
	};

	return (
		<Controller
			name={name}
			control={control}
			render={({ field, fieldState: { error } }) => {
				const handleOnChange = (event : React.ChangeEvent<HTMLInputElement>) => {
					onChange && onChange(event);
					field.onChange(event);
				};
				return (
					<PasswordInput
						{...field}
						data-testid="password"
						ref={ref}
						error={!!error}
						autoComplete="off"
						onChange={handleOnChange}
						icon={<OpenedEyeOutline fontSize={14} />}
						visibilityToggleIcon={getVisibilityIcon}
						label={error?.message ? error.message : label}
						classNames={{
							icon       : classes.icon,
							input      : classes.pswInput,
							innerInput : classes.innerInput,
							invalid    : classes.invalidPassword,
							wrapper    : error && classes.invalidWrapper,
							label      : `${classes.label} ${error && classes.labelError}`,
						}}
						{...rest}
					/>
				);
			}}
		/>
	);
};

export default RHFPasswordInput;
