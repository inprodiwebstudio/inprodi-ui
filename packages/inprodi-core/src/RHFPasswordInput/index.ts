export { RHFPasswordInput } from "./RHFPasswordInput";
export type { RHFPasswordInputProps } from "./RHFPasswordInput";