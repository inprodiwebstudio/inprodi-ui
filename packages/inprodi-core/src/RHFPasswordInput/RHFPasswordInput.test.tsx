import React                      from "react";
import * as Yup                   from "yup";
import { FormProvider, useForm }  from "react-hook-form";
import { yupResolver }            from "@hookform/resolvers/yup";
import { act, fireEvent, render } from "@testing-library/react";
import RHFPasswordInput           from "./RHFPasswordInput";

describe("RHFPasswordInput.test", () => {
	test("snapshot", () => {
		const textInptName = "password";

		  const Provider: React.FC<{children : React.ReactElement }> = ({ children }) => {
			const methods = useForm<{ [textInptName]: string }>({ defaultValues : {
				[textInptName] : "",
			}});
			return <FormProvider {...methods}>{children}</FormProvider>;
		};

		const { container } = render(
			<Provider>
				<RHFPasswordInput name={textInptName} />
			</Provider>
		);

		expect(container).toMatchSnapshot();
	});
	test("should change value on RHK and call the onChange prop and the onChange callback must receive the HTMLInputElement as a arg", () => {
		const inptName = "password";
		const value = "testVal";
		let customVal = "";

		const Provider: React.FC<{children : React.ReactElement }> = ({ children }) => {
			const methods = useForm<{ [inptName]: string }>({ defaultValues : {
				[inptName] : "",
			}});
			return <FormProvider {...methods}>{children}</FormProvider>;
		};

		const hadleOnChage = (e : React.ChangeEvent<HTMLInputElement>) => {
			customVal = `hadleOnChage${e.target.value}`;
		};

		const { getByTestId, getByDisplayValue } = render(
			<Provider>
				<RHFPasswordInput name={inptName} onChange={hadleOnChage} />
			</Provider>
		);

		const inpt = getByTestId("password");
		fireEvent.change(inpt, { target : { value : value } });

		expect(getByDisplayValue(value).getAttribute("name")).toBe(inptName);
		expect(customVal).toBe(`hadleOnChage${value}`);
	});
	test("should has the ...rest props", () => {
		const Provider: React.FC<{children : React.ReactElement }> = ({ children }) => {
			const methods = useForm<{ test : string }>();
			return <FormProvider {...methods}>{children}</FormProvider>;
		};

		const { getByTestId } = render(
			<Provider>
				<RHFPasswordInput name="test" disabled />
			</Provider>
		);

		const inpt = getByTestId("password");

		expect(inpt.hasAttribute("disabled")).toBeTruthy();
	});
	test("should to show the correct labels, when there is some error the label should to change to error label", async () => {
		const onSubmit = jest.fn();
		const labelError = "Error";

		const Provider: React.FC<{children : React.ReactElement }> = ({ children }) => {
			const schema = Yup.object().shape({
				test : Yup.string()
					.required(labelError),
			});
			const methods = useForm<{ test : string }>({
				resolver : yupResolver(schema),
			});

			return (
				<FormProvider {...methods} handleSubmit={onSubmit}>
					 <form onSubmit={methods.handleSubmit(onSubmit)}>
						{children}
						<button type="submit">SubmitBtn</button>
					 </form>
				</FormProvider>
			);
		};

		const { getByRole, getByLabelText } = render(
			<Provider>
				<RHFPasswordInput label="labelTest" name="test" />
			</Provider>
		);

		expect(getByLabelText("labelTest")).toBeTruthy();

		const btn = getByRole("button");
		await act( () => fireEvent.submit(btn));


		expect(getByLabelText(labelError)).toBeTruthy();
	});
	test("should has the error styles", async () => {
		const onSubmit = jest.fn();

		const Provider: React.FC<{children : React.ReactElement }> = ({ children }) => {
			const schema = Yup.object().shape({
				test : Yup.string()
					.required("Error"),
			});
			const methods = useForm<{ test : string }>({
				resolver : yupResolver(schema),
			});

			return (
				<FormProvider {...methods} handleSubmit={onSubmit}>
					 <form onSubmit={methods.handleSubmit(onSubmit)}>
						{children}
						<button type="submit">SubmitBtn</button>
					 </form>
				</FormProvider>
			);
		};

		const { getByRole, container } = render(
			<Provider>
				<RHFPasswordInput label="labelTest" name="test" />
			</Provider>
		);

		const inpt = container.querySelector("input");
		const btn = getByRole("button");

		await act( () => fireEvent.submit(btn));

		expect(inpt?.className.includes("mantine-PasswordInput-invalid")).toBeTruthy();
	});
});
