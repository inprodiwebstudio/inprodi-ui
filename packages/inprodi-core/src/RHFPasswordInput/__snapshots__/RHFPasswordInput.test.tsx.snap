// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RHFPasswordInput.test snapshot 1`] = `
<div>
  <div
    class="mantine-InputWrapper-root mantine-PasswordInput-root mantine-1m3pqry"
  >
    <div
      class="mantine-Input-wrapper mantine-PasswordInput-wrapper mantine-12sbrde"
    >
      <div
        class="__mantine-ref-icon mantine-Input-icon mantine-PasswordInput-icon mantine-1qxjidu"
      >
        <svg
          fill="none"
          font-size="14"
          height="16"
          viewBox="0 0 14 15"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.996 2.774C3.694 2.774.786 5.062.014 8.06a.438.438 0 1 0 .847.217c.665-2.584 3.242-4.63 6.135-4.63s5.478 2.047 6.143 4.63a.438.438 0 0 0 .847-.217c-.772-3-3.689-5.287-6.99-5.287Zm.005 2.333a2.825 2.825 0 0 0-2.82 2.819 2.826 2.826 0 0 0 2.82 2.819 2.826 2.826 0 0 0 2.819-2.82A2.826 2.826 0 0 0 7 5.107Zm0 .875c1.078 0 1.944.865 1.944 1.944A1.938 1.938 0 0 1 7 9.87a1.937 1.937 0 0 1-1.944-1.944c0-1.08.865-1.944 1.944-1.944Z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        aria-invalid="false"
        class="mantine-PasswordInput-input mantine-Input-input mantine-PasswordInput-input mantine-Input-withIcon mantine-PasswordInput-withIcon mantine-1qyinf5"
      >
        <input
          autocomplete="off"
          class="__mantine-ref-innerInput mantine-PasswordInput-innerInput mantine-PasswordInput-withIcon mantine-9rl397"
          data-testid="password"
          id="mantine-r0"
          name="password"
          type="password"
          value=""
        />
      </div>
      <div
        class="mantine-1jdmi3h mantine-Input-rightSection mantine-PasswordInput-rightSection"
      >
        <button
          aria-hidden="true"
          class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-PasswordInput-visibilityToggle mantine-14z3xly"
          tabindex="-1"
          type="button"
        >
          <svg
            data-testid="closed-eye"
            fill="none"
            font-size="15"
            height="16"
            viewBox="0 0 15 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.993.774C4.304.774 1.827 1.98.143 3.846a.552.552 0 0 0 .82.74c1.481-1.64 3.673-2.708 6.03-2.708 2.356 0 4.552 1.069 6.038 2.71a.554.554 0 0 0 .935-.557.553.553 0 0 0-.117-.185C12.161 1.98 9.681.774 6.993.774Zm6.034 6.619a.552.552 0 0 0-.508.418c-.59 2.217-2.916 4.009-5.526 4.009s-4.929-1.79-5.52-4.009a.552.552 0 1 0-1.067.285c.176.663.474 1.282.86 1.846l-1.049.874a.553.553 0 1 0 .708.848l1.101-.918c.4.413.83.776 1.328 1.082l-.675 1.17a.552.552 0 1 0 .957.552l.7-1.212c.667.277 1.361.482 2.108.541v1.334a.552.552 0 1 0 1.105 0V12.88c.745-.06 1.44-.266 2.107-.542l.7 1.213a.552.552 0 1 0 .957-.552l-.676-1.172a6.739 6.739 0 0 0 1.328-1.081l1.102.919a.554.554 0 1 0 .708-.848l-1.05-.875a6.016 6.016 0 0 0 .862-1.845.553.553 0 0 0-.56-.703Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
